internal/graph/lookupsubjects.go

```diff
@@ -495,114 +905,389 @@ func (lsu *lookupSubjectsUnion) CompletedChildOperations() error {
+		parentCi:     ci,
+		combinationHandler: func(fs datasets.SubjectSetByResourceID, other datasets.SubjectSetByResourceID) error {
+			fs.SubtractAll(other)
+			return nil
+		},
+		firstBranchCi: ci,
 	}
 }
 
-func (lse *lookupSubjectsExclusion) ForIndex(ctx context.Context, setOperationIndex int) dispatch.LookupSubjectsStream {
-	collector := dispatch.NewCollectingDispatchStream[*v1.DispatchLookupSubjectsResponse](ctx)
-	lse.collectors[setOperationIndex] = collector
-	return collector
+// finalSubjectIDForResults returns the ID of the last subject (sorted) in the results, if any.
+// Returns empty string if none.
+func finalSubjectIDForResults(ci cursorInformation, results []*v1.DispatchLookupSubjectsResponse) (string, error) {
+	endingSubjectIDs := mapz.NewSet[string]()
+	for _, result := range results {
+		frc, err := newCursorInformation(result.AfterResponseCursor, ci.limits, lsDispatchVersion)
+		if err != nil {
+			return "", err
+		}
+
+		lastSubjectID, _ := frc.headSectionValue()
+		if lastSubjectID == "" {
+			return "", spiceerrors.MustBugf("got invalid cursor")
+		}
+
+		endingSubjectIDs.Add(lastSubjectID)
+	}
+
+	sortedSubjectIDs := endingSubjectIDs.AsSlice()
+	sort.Strings(sortedSubjectIDs)
+
+	if len(sortedSubjectIDs) == 0 {
+		return "", nil
+	}
+
+	return sortedSubjectIDs[len(sortedSubjectIDs)-1], nil
 }
 
-func (lse *lookupSubjectsExclusion) CompletedChildOperations() error {
-	var foundSubjects datasets.SubjectSetByResourceID
-	metadata := emptyMetadata
+// createFilteredAndLimitedResponse creates a filtered and limited (as is necessary via the cursor and limits)
+// version of the subjects, returning a DispatchLookupSubjectsResponse ready for publishing with just that
+// subset of results.
+func createFilteredAndLimitedResponse(
+	ci cursorInformation,
+	subjects map[string]*v1.FoundSubjects,
+	metadata *v1.ResponseMeta,
+) (*v1.DispatchLookupSubjectsResponse, func(), error) {
+	if subjects == nil {
+		return nil, func() {}, spiceerrors.MustBugf("nil subjects given to createFilteredAndLimitedResponse")
+	}
 
-	for index := 0; index < len(lse.collectors); index++ {
-		collector := lse.collectors[index]
-		results := datasets.NewSubjectSetByResourceID()
-		for _, result := range collector.Results() {
-			metadata = combineResponseMetadata(metadata, result.Metadata)
-			if err := results.UnionWith(result.FoundSubjectsByResourceId); err != nil {
-				return fmt.Errorf("failed to UnionWith under lookupSubjectsExclusion: %w", err)
+	afterSubjectID, _ := ci.headSectionValue()
+
+	// Filter down the subjects found by the cursor (if applicable) and the apply a limit.
+	filteredSubjectIDs := mapz.NewSet[string]()
+	for _, foundSubjects := range subjects {
+		for _, foundSubject := range foundSubjects.FoundSubjects {
+			// NOTE: wildcard is always returned, because it is needed by all branches, at all times.
+			if foundSubject.SubjectId == tuple.PublicWildcard || (afterSubjectID == "" || foundSubject.SubjectId > afterSubjectID) {
```

Comment:
in not sure to understand how we guarantee that we do not skip found subjects that are alphabetically after the current cursor but that hasn't been seen by the stream before.

Reply:
Because the stream always returns in sorted order from the root. That way, we are always guaranteed to have a defined ordering (alphabetical) coming out of each subproblem